cub/cub/device/dispatch/dispatch_transform.cuh

```diff
@@ -0,0 +1,1128 @@
+  {
+    smem_offset = round_up_to_po2_multiple(smem_offset, static_cast<int>(alignof(T)));
+  }
+  auto smem_dst = reinterpret_cast<T*>(smem + smem_offset);
+  assert(reinterpret_cast<uintptr_t>(smem_dst) % memcpy_async_size_multiple == 0); // to hit optimal memcpy_async
+                                                                                   // performance
+  cooperative_groups::memcpy_async(
+    group,
+    smem_dst,
+    aligned_ptr.ptr + global_offset,
+    aligned_size_t<memcpy_async_size_multiple>{static_cast<::cuda::std::size_t>(count)});
+  smem_offset += count;
+  return smem_dst + aligned_ptr.offset;
+}
+
+template <typename, typename Offset, typename F, typename RandomAccessIteartorOut, typename... InTs>
+_CCCL_DEVICE void transform_kernel_impl(
+  ::cuda::std::integral_constant<Algorithm, Algorithm::memcpy_async>,
+  Offset num_items,
+  int num_elem_per_thread,
+  F f,
+  RandomAccessIteartorOut out,
+  aligned_base_ptr<const InTs>... aligned_ptrs)
+{
+  extern __shared__ char smem[]; // this should be __attribute((aligned(memcpy_async_alignment))), but then it clashes
+                                 // with the ublkcp kernel, which sets a higher alignment, since they are both called
+                                 // from the same kernel entry point (albeit one is always discarded). However, SMEM is
+                                 // 16-byte aligned by default.
+
+  const Offset tile_stride   = blockDim.x * num_elem_per_thread;
+  const Offset global_offset = static_cast<Offset>(blockIdx.x) * tile_stride;
+  const int tile_size        = ::cuda::std::min(num_items - global_offset, tile_stride);
+
+  auto group           = cooperative_groups::this_thread_block();
+  int smem_offset      = 0;
+  const auto smem_ptrs = ::cuda::std::tuple<const InTs*...>{
+    copy_and_return_smem_dst(group, tile_size, smem, smem_offset, global_offset, aligned_ptrs)...};
+  cooperative_groups::wait(group);
+  (void) smem_ptrs; // suppress unused warning for MSVC
+  (void) &smem_offset; // MSVC needs extra strong unused warning supression
+
+#pragma unroll 1
+  for (int i = 0; i < num_elem_per_thread; ++i)
+  {
+    const int smem_idx    = i * blockDim.x + threadIdx.x;
+    const Offset gmem_idx = global_offset + smem_idx;
+    if (gmem_idx < num_items)
+    {
+      out[gmem_idx] = poor_apply(
+        [&](const InTs*... smem_base_ptrs) {
+          return f(smem_base_ptrs[smem_idx]...);
+        },
+        smem_ptrs);
+    }
+  }
+}
+
+constexpr int ublkcp_alignment     = 128;
+constexpr int ublkcp_size_multiple = 16;
+
+#ifdef _CUB_HAS_TRANSFORM_UBLKCP
+// TODO(bgruber): inline this as lambda in C++14
+template <typename T>
+_CCCL_DEVICE void copy_ptr_set(
+  uint64_t& bar,
+  uint32_t tile_size,
+  int tile_stride,
+  char* smem,
+  int& smem_offset,
+  uint32_t& total_copied,
+  size_t global_offset,
+  const aligned_base_ptr<T>& aligned_ptr)
+{
+#  if CUB_PTX_ARCH >= 900
```

Comment:
This is not as simple as it sounds, as the PTX arch is mostly not dependent on the architecture the code is running on, but the Version of NVCC the code was compiled with

Reply:
I am slightly confused by @miscco's explanation. The intend here is to remove code that cannot be compiled below 900, because the PTX intrinsics/mneumonics are not available. E.g. `ptx::cp_async_bulk` cannot be compiled  below 900. There is no fallback path.

Would you still prefer to use `NV_IF_TARGET`?

Reply:
So with "PTX arch" you ment the "PTX ISA version"? That would make sense to me.

> You can build this code just fine with a recent nvcc but run it on old hardware.

I don' think so. The instruction requires at least Hopper. Compiling for anything lower results in a linker error.

> So this is not a "replace with NV_IF_TARGET" but at best a "additionally use NV_IF_TARGET to avoid unsupported architectures when compiling with modern nvcc"

I could technically replace the `#  if CUB_PTX_ARCH >= 900` with `NV_IF_TARGET` though.  I don't see why I would need both checks.

If it matters, `_CUB_HAS_TRANSFORM_UBLKCP` is only defined for nvcc >= 12.